### PR TITLE
#48buyer should apply coupon during checkout

### PR DESCRIPTION
**Pull request by Luciefifi**
# What does this PR do
This PR is for applying coupons by buyer

# Description of the task to be completed
as a buyer , I should be able to apply coupons while checking out 

# how should this be manually tested
. to test this start by cloning this repository
. cd into it
. checkout to  ft-apply-coupon
. run migrations by `npm run migrate:up`
. run seedes by `npm run db :seeds`
. start the server by `npm run start`
. then make a post request to apply coupon

# Relevant pivotal tracker user story
#55

### CODE REVIEW

**Review by sostene12 — APPROVED** [review]

**Review by patsicko — APPROVED** [review]

File: src/controllers/couponCodeController.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+import generateCouponCode from "../utils/helpers/couponGenerate";
+
+const { CouponCodeDiscount, Product, Vendor, Buyer, Cart } = models;
+
+class CouponCodeController {
+  static async createCouponCode(req, res) {
+    const { productId, discount, expirationDate, maxUsage } = req.body;
+    if (discount <= 0) {
+      return res
+        .status(409)
```
Comment (wakaflorien) [inline src/controllers/couponCodeController.js]: - i think the correct status code for 0/no discount is just 400 not 409 for conflict.

File: src/controllers/couponCodeController.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+      });
+      return res.status(200).json({ status: "succcess", data: vendorCoupons });
+    } catch (error) {
+      return res.status(500).json({ status: "fail", error: error.message });
+    }
+  }
+
+  static async checkExpiredCoupons(req, res) {
+    try {
+      const expiredCoupons = await CouponCodeDiscount.findAll({
```
Comment (wakaflorien) [inline src/controllers/couponCodeController.js]: - the result of this is clearly not expired couponCodes.
- i suggest you check this in condition for truthie on not then after you proceed by keeping the discount status as it is on setting it to true.

File: src/controllers/couponCodeController.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      const vendorCoupons = await CouponCodeDiscount.findAll({
+        where: { VendorId: vendor.dataValues.id }
+      });
+      return res.status(200).json({ status: "succcess", data: vendorCoupons });
+    } catch (error) {
+      return res.status(500).json({ status: "fail", error: error.message });
+    }
+  }
+
+  static async checkExpiredCoupons(req, res) {
```
Comment (wakaflorien) [inline src/controllers/couponCodeController.js]: - the function name `checkExpiredCoupons` gives the image of oly checking while you are doing more in it which is disabling discount
- you can seperate the two may or keep it in one function and you change the name.

File: src/controllers/couponCodeController.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+          .json({ status: "fail", message: "Maximum usage reached" });
+      }
+      let eligibleProducts = [];
+      const product = await Product.findByPk(coupon.dataValues.product);
+      if (!product) {
+        return res.status(400).json({ message: "Invalid product ID" });
+      }
+      eligibleProducts.push(product.dataValues);
+      const cart = await Cart.findOne({ where: { buyerId: req.user.id } });
+      console.log(buyer);
```
Comment (wakaflorien) [inline src/controllers/couponCodeController.js]: - you can remove this console when you are done.

Comment (Luciefifi) [reply src/controllers/couponCodeController.js]: Thank you so much @wakaflorien  for your review , let me work on them

**Review by chrill-christella — APPROVED** [review]

**Review by wakaflorien — APPROVED** [review]

**Pull request merged**